backend/pikpak/pikpak.go

```diff
@@ -0,0 +1,1749 @@
+		// delete it.
+		err = f.deleteObjects(ctx, []string{rootID}, trashedFiles || f.opt.UseTrash)
+		if err != nil {
+			return err
+		}
+	} else if check {
+		return errors.New("can't purge root directory")
+	}
+	f.dirCache.FlushDir(dir)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Rmdir deletes the root folder
+//
+// Returns an error if it isn't empty
+func (f *Fs) Rmdir(ctx context.Context, dir string) error {
+	return f.purgeCheck(ctx, dir, true)
+}
+
+// Purge deletes all the files and the container
+//
+// Optional interface: Only implement this if you have a way of
+// deleting all the files quicker than just running Remove() on the
+// result of List()
+func (f *Fs) Purge(ctx context.Context, dir string) error {
+	return f.purgeCheck(ctx, dir, false)
+}
+
+// CleanUp empties the trash
+func (f *Fs) CleanUp(ctx context.Context) (err error) {
+	opts := rest.Opts{
+		Method:     "PATCH",
+		Path:       "/drive/v1/files/trash:empty",
+		NoResponse: true, // Only returns `{"task_id":""}
+	}
+	var resp *http.Response
+	err = f.pacer.Call(func() (bool, error) {
+		resp, err = f.srv.Call(ctx, &opts)
+		return f.shouldRetry(ctx, resp, err)
+	})
+	if err != nil {
+		return fmt.Errorf("couldn't empty trash: %w", err)
+	}
+	return nil
+}
+
+// Move the object
+func (f *Fs) moveObjects(ctx context.Context, IDs []string, dirID string) (err error) {
+	if len(IDs) == 0 {
+		return nil
+	}
+	req := api.RequestBatch{
+		Ids: IDs,
+		To:  &map[string]string{"parent_id": parentIDForRequest(dirID)},
+	}
+	if err := f.requestBatchAction(ctx, "batchMove", &req); err != nil {
+		return fmt.Errorf("move object failed: %w", err)
+	}
+	return nil
+}
+
+// renames the object
+func (f *Fs) renameObject(ctx context.Context, ID, newName string) (info *api.File, err error) {
+	req := api.File{
+		Name: f.opt.Enc.FromStandardName(newName),
+	}
+	info, err = f.patchFile(ctx, ID, &req)
+	if err != nil {
+		return nil, fmt.Errorf("rename object failed: %w", err)
+	}
+	return
+}
+
+// DirMove moves src, srcRemote to this remote at dstRemote
+// using server-side move operations.
+//
+// Will only be called if src.Fs().Name() == f.Name()
+//
+// If it isn't possible then return fs.ErrorCantDirMove
+//
+// If destination exists then return fs.ErrorDirExists
+func (f *Fs) DirMove(ctx context.Context, src fs.Fs, srcRemote, dstRemote string) error {
+	srcFs, ok := src.(*Fs)
+	if !ok {
+		fs.Debugf(srcFs, "Can't move directory - not same remote type")
+		return fs.ErrorCantDirMove
+	}
+
+	srcID, srcParentID, srcLeaf, dstParentID, dstLeaf, err := f.dirCache.DirMove(ctx, srcFs.dirCache, srcFs.root, srcRemote, f.root, dstRemote)
+	if err != nil {
+		return err
+	}
+
+	if srcParentID != dstParentID {
+		// Do the move
+		err = f.moveObjects(ctx, []string{srcID}, dstParentID)
+		if err != nil {
+			return fmt.Errorf("couldn't dir move: %w", err)
+		}
+	}
+
+	// Can't copy and change name in one step so we have to check if we have
```

Comment:
This can cause data loss if there was an existing object with the old name at the new position.

Is that true?

I thought the unit tests tested for this...

Reply:
OK great!